cloudbridge/cloud/providers/aws/resources.py

```diff
@@ -652,79 +560,71 @@ def __init__(self, provider, security_group):
-                    else None):
-                return self.get_rule(ip_protocol, from_port, to_port, cidr_ip,
-                                     src_group)
-        except EC2ResponseError as ec2e:
-            if ec2e.code == "InvalidPermission.Duplicate":
+            src_group_id = (
+                src_group.id if isinstance(src_group, SecurityGroup)
+                else src_group)
+
+            ip_perm_entry = {
+                'IpProtocol': ip_protocol,
+                'FromPort': from_port,
+                'ToPort': to_port,
+                'IpRanges': [{'CidrIp': cidr_ip}] if cidr_ip else None,
+                'UserIdGroupPairs': [{
+                    'GroupId': src_group_id}
+                ] if src_group_id else None
+            }
+            # Filter out empty values to please Boto
+            ip_perms = [trim_empty_params(ip_perm_entry)]
+            self._security_group.authorize_ingress(IpPermissions=ip_perms)
+            self._security_group.reload()
+            return self.get_rule(ip_protocol, from_port, to_port, cidr_ip,
+                                 src_group_id)
+        except ClientError as ec2e:
+            if ec2e.response['Error']['Code'] == "InvalidPermission.Duplicate":
                 return self.get_rule(ip_protocol, from_port, to_port, cidr_ip,
                                      src_group)
             else:
                 raise ec2e
-        return None
 
     def get_rule(self, ip_protocol=None, from_port=None, to_port=None,
                  cidr_ip=None, src_group=None):
-        for rule in self._security_group.rules:
-            if (rule.ip_protocol == ip_protocol and
-                rule.from_port == from_port and
-                rule.to_port == to_port and
-                rule.grants[0].cidr_ip == cidr_ip) or \
-                    (rule.grants[0].group_id == src_group.id if src_group and
-                        hasattr(rule.grants[0], 'group_id') else False):
-                return AWSSecurityGroupRule(self._provider, rule, self)
+        src_group_id = (src_group.id if isinstance(src_group, SecurityGroup)
+                        else src_group)
+        for rule in self._security_group.ip_permissions:
+            if ip_protocol and rule['IpProtocol'] != ip_protocol:
+                continue
+            elif from_port and rule['FromPort'] != from_port:
+                continue
+            elif to_port and rule['ToPort'] != to_port:
+                continue
+            elif cidr_ip:
+                if cidr_ip not in [x['CidrIp'] for x in rule['IpRanges']]:
+                    continue
+            elif src_group_id:
+                if src_group_id not in [
+                    group_pair.get('GroupId') for group_pair in
+                        rule.get('UserIdGroupPairs', [])]:
+                    continue
+            return AWSSecurityGroupRule(self._provider, rule, self)
         return None
 
     def to_json(self):
```

Comment:
We could use the ec2 client `describe_security_groups` method here directly and not worry about translating this into json: http://boto3.readthedocs.io/en/stable/reference/services/ec2.html?highlight=create_volume#EC2.Client.describe_security_groups:
`return self._provider.ec2_conn.meta.client.describe_security_group(GroupIds=[self.id])`

Reply:
Can you elaborate - I didn't quite follow. I'm hoping this part can be redone anyway when we do the interface consistency change for security groups rules, so that might be a good time to address this.

Reply:
It looks like `boto3` client has the `describe` method for various objects that return the native native object representation so rather than providing our own implementation like we had to with `boto`, it should be easier to just call the native `boto3` method. This is an example of what the describe method will return:
```
[{'Description': 'default group',
   'GroupId': 'sg-7d10cc15',
   'GroupName': 'default',
   'IpPermissions': [{'FromPort': 0,
     'IpProtocol': 'tcp',
     'IpRanges': [],
     'Ipv6Ranges': [],
     'PrefixListIds': [],
     'ToPort': 65535,
     'UserIdGroupPairs': [{'GroupId': 'sg-7d10cc15',
       'GroupName': 'default',
       'UserId': '347162595075'}]},
    {'FromPort': 0,
     'IpProtocol': 'udp',
     'IpRanges': [],
     'Ipv6Ranges': [],
     'PrefixListIds': [],
     'ToPort': 65535,
     'UserIdGroupPairs': [{'GroupId': 'sg-7d10dc15',
       'GroupName': 'default',
       'UserId': '347162595075'}]},
    {'FromPort': -1,
     'IpProtocol': 'icmp',
     'IpRanges': [],
     'Ipv6Ranges': [],
     'PrefixListIds': [],
     'ToPort': -1,
     'UserIdGroupPairs': [{'GroupId': 'sg-7d10dc15',
       'GroupName': 'default',
       'UserId': '9999999999'}]}],
   'IpPermissionsEgress': [],
   'OwnerId': '999999999'}]
```
Granted, it's dict but we could just return `json.loads` of that?

Reply:
So you're saying we should make a server call instead of using the cached ip_permissions value?